File: dj_link.py
import requests
import re
import json
from urllib.parse import urlparse

"""
提取东京优惠券api
"""

url = input("请输入URL: ")  # 用户输入URL

parsed_url = urlparse(url)
host = parsed_url.netloc  # 提取URL中的Host部分

headers = {
    'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 16_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Mobile/15E148 Safari/604.1',
    'Host': host,
}

response = requests.get(url, headers=headers)
response.encoding = response.apparent_encoding  # 尝试自动识别编码
html = response.text

regex = r'window\.__api_data__\s*=\s*({[^;]+});'  # 匹配 window.__api_data__ = {...};
match = re.search(regex, html)

if match and match.group(1):
    json_data = match.group(1)
    try:
        api_data = json.loads(json_data)
        if api_data and 'floorList' in api_data:
            for floor in api_data['floorList']:
                if 'couponList' in floor:
                    for coupon in floor['couponList']:
                        print('beginTimeStr:', coupon.get('beginTimeStr'))
                        print('limit:', coupon.get('limit'))
                        print('args:', coupon.get('args'))
                        print('encodeActivityId:', api_data.get('encodeActivityId'))
                        print('scene:', coupon.get('scene'))
                        print('-------------')
                        # 在这里进行你需要的处理
    except json.JSONDecodeError as e:
        print('JSON 解析错误：', e)
    else:
        print('没有找到符合条件的 JSON 数据。')

File: dj_link.js
const axios = require('axios');
const cheerio = require('cheerio');
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// 通过 readline 提问用户输入 targetUrl
rl.question('Please enter the target URL: ', (targetUrl) => {
  rl.close();

  const headers = {
    'User-Agent': 'Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36',
    'Host': 'prodev.m.jd.com'
  };

  axios.get(targetUrl, { headers })
    .then(response => {
      const $ = cheerio.load(response.data);

      // 在所有的 script 标签中查找包含 window.__api_data__ 的内容
      $('script').each((index, element) => {
        const scriptContent = $(element).html();
        if (scriptContent.includes('window.__api_data__')) {
          // 使用正则表达式提取 window.__api_data__ 内容
          const match = scriptContent.match(/window\.__api_data__\s*=\s*({.*?});/);
          if (match && match[1]) {
            const jsonString = match[1];

            // 尝试解析 JSON 数据
            try {
              const jsonData = JSON.parse(jsonString);

              // 从 targetUrl 中提取 activityId
              const activityIdMatch = targetUrl.match(/\/active\/([^\/]+)/);
              const activityId = activityIdMatch && activityIdMatch[1];

              // 构建新的 URL
              if (jsonData.floorList && jsonData.floorList.length > 0) {
                jsonData.floorList.forEach(floor => {
                  if (floor.couponList && floor.couponList.length > 0) {
                    floor.couponList.forEach(coupon => {
                      const channel = coupon.channel;
                      const scene = coupon.scene;
                      const args = coupon.args;

                      // 构建新的 URL
                      const newUrl = `https://api.m.jd.com/client.action?functionId=newBabelAwardCollection&body={"activityId":"${activityId}","scene":"${scene}","args":"${args}"`;

                      // 输出新的 URL
                      console.log('New URL:', newUrl);
                      console.log('-------------------------');
                    });
                  }
                });
              }
            } catch (error) {
              console.error('Error parsing JSON:', error);
            }
          }
        }
      });
    })
    .catch(error => {
      console.error('Error fetching the page:', error);
    });
});
